src/Spice86.Core/Emulator/Devices/Video/Renderer.cs

```diff
@@ -54,11 +82,32 @@ public void Render(Span<uint> frameBuffer) {
         }
         try {
             BufferSize = frameBuffer.Length;
-            if (Width * Height > BufferSize) {
+            
+            // Calculate native (pre-aspect-correction) dimensions
+            int nativeHeight = CalculateNativeHeight();
+            int width = Width;
+            int outputHeight = Height;
+            bool needsAspectCorrection = outputHeight != nativeHeight;
+            
+            if (width * outputHeight > BufferSize) {
                 // Resolution change hasn't caught up yet. Skip a frame.
                 return;
             }
 
+            // If aspect correction is needed, render to temp buffer first
+            Span<uint> renderTarget;
```

Comment:
reduce cyclomatic complexity.

Keep a maximum of changes outside of the renderer, so they can be unit tested with, and secondly minimize complexity. Prefer Pure funections.